### DISCUSSION TITLE
How to use Java workflow

### DISCUSSION THREAD

**Discussion by Chris-playground** [apache/dolphinscheduler#10769] [Q&A]
We use a java program in Oozie to finish transporting data from hdfs to Doris.
But I couldn't find a workflow type of Java Program, then I use MR and choose JAVA type, but I had few questions that needs your kind help.
- how to use libjars, i noticed that you have a `-libjars` param in the manual, but it doesn't have an example. Should I use `-libjars=path` and upload java dependencies in this directory or use `-libjars path`? i tried both and they all failed.
- my java program use arg4j to pass params. In Oozie we config like this
```java
-loadLabelUrl
/user/bdads/shell-doris/load_dim_sap_plant.sql(HDFS path)
-maxWaitMinutes
2
-everyWaitingMinutes
0.1
```
but I don't know how to config in DS, I tried severl ways but failed.
- what is the differenct between main arguments and option parameters?

**Comment by Chris-playground** [comment]
I found that the manual has quite a different descriptions with the new version, could you please update the docs when you publish a new release?

**Comment by EricGao888** [comment]
Hi @Chris-Arith, currently there is a PR working on `java task plugin`. BTW, why not try `shell task plugin` to run your java program?

FYI https://github.com/apache/dolphinscheduler/issues/10683

**Reply by Chris-playground** [reply]
Thanks, I have already tried the shell task plugin, it worked, but I have a problem here.
I have to package the jar with dependencies and it's big, is there any way that i could use something like sharelib in Oozie that i only need to upload a jar without dependencies.

**Reply by EricGao888** [reply]
What about uploading a so-called `sharelib` into ds resource center and choosing both the `sharelib` and your java `jar` in your shell task? However, there could be performance issue since workers need to download the resource from storage. May I ask whether there are better approaches to handle this? @zhongjiajie @SbloodyS @caishunfeng

**Reply by zhongjiajie** [reply]
Maybe @EricGao888 is better way, but you could also add sharelib to all work, like in path `/path/to/share`, and execute the jar package with add  `/path/to/share` to your classpath

**Reply by EricGao888** [reply]
> Maybe @EricGao888 is better way, but you could also add sharelib to all work, like in path `/path/to/share`, and execute the jar package with add `/path/to/share` to your classpath

Cool, adding sharelib to all workers could avoid performance issues caused by duplicate resource download.

**Reply by Chris-playground** [reply]
how about using git to manage these sharelib

**Reply by EricGao888** [reply]
> how about using git to manage these sharelib

We do not recommend to use git protocol to manage large files like sharelib. You may refer to comments here: https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue+is%3Aopen+git#issuecomment-1166904625

To add source control for large files, maybe you could consider implementing some source control mechanism based on object storage like AWS S3, Alibaba Cloud OSS, etc.

**Reply by Chris-playground** [reply]
> > how about using git to manage these sharelib
> 
> We do not recommend to use git protocol to manage large files like sharelib. You may refer to comments here: https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue+is%3Aopen+git#issuecomment-1166904625
> 
> To add source control for large files, maybe you could consider implementing some source control mechanism based on object storage like AWS S3, Alibaba Cloud OSS, etc.

ok, thanks

**Reply by zhongjiajie** [reply]
> > Maybe @EricGao888 is better way, but you could also add sharelib to all work, like in path `/path/to/share`, and execute the jar package with add `/path/to/share` to your classpath
> 
> Cool, adding sharelib to all workers could avoid performance issues caused by duplicate resource download.

This method has obvious disadvantages, you can not manager them, if you want to add/delete them you should change by hand or extra script

**Reply by Chris-playground** [reply]
I have come up with another problem that I couldn't connnect to impala, could you help me please..

**Reply by Chris-playground** [reply]
> I have come up with another problem that I couldn't connnect to impala, could you help me please..

I have already added ImpalaJDBC41.jar into api-server/libs, but when i click test connection, the request was cancelled and have no reponse

**Reply by zhongjiajie** [reply]
you should restart the api server when before you test it

**Reply by Chris-playground** [reply]
> you should restart the api server when before you test it

I used the install.sh to restart the whole DS project, it didn't work.

**Reply by EricGao888** [reply]
@Chris-Arith BTW, I suggest you joining the Apache DolphinScheduler Slack channel, it is a better place for Q & A.

**Reply by Chris-playground** [reply]
> @Chris-Arith BTW, I suggest you joining the Apache DolphinScheduler Slack channel, it is a better place for Q & A.

tks, let me check

**Reply by zhongjiajie** [reply]
> 

you should not rerun install(not sure whch step you add new jar to api-server/liibs), you should better run command `sh ./bin/dolphinscheduler-daemon.sh stop api-server; sh ./bin/dolphinscheduler-daemon.sh stop api-server` after you add new package. see more detail in https://dolphinscheduler.apache.org/en-us/docs/latest/user_doc/guide/installation/pseudo-cluster.html section "Start or Stop Server"

**Reply by Chris-playground** [reply]
> > 
> 
> you should not rerun install(not sure whch step you add new jar to api-server/liibs), you should better run command `sh ./bin/dolphinscheduler-daemon.sh stop api-server; sh ./bin/dolphinscheduler-daemon.sh stop api-server` after you add new package. see more detail in https://dolphinscheduler.apache.org/en-us/docs/latest/user_doc/guide/installation/pseudo-cluster.html section "Start or Stop Server"

Firstly, I tried to restart api-server in this way and it didn't work also.
I used the `install.sh` to add new jars and update config to cluster members and I found it work, maybe it is unstable using this way , I will stop trying like this next time.

**Reply by Chris-playground** [reply]
> @Chris-Arith BTW, I suggest you joining the Apache DolphinScheduler Slack channel, it is a better place for Q & A.

The slack invitation link is no longer valid, could you please send me another invitation?

**Reply by EricGao888** [reply]
@Chris-Arith Please try this one https://join.slack.com/t/asf-dolphinscheduler/shared_invite/zt-l8k90ceu-wwUfobaDkJxjzMfZp4y1Ag

**Reply by Chris-playground** [reply]
> @Chris-Arith Please try this one https://join.slack.com/t/asf-dolphinscheduler/shared_invite/zt-l8k90ceu-wwUfobaDkJxjzMfZp4y1Ag

![image](https://user-images.githubusercontent.com/48941586/177510856-f1bd06c5-1c35-486f-a864-82ce069566c2.png)
